### PR TITLE
Node Drain API and CLI

### PR DESCRIPTION
**Pull request by dadgar**
This PR introduces the new node drain API and exposes it via the CLI command.

### CODE REVIEW

File: command/agent/node_endpoint.go [hunk 1]
```diff
@@ -101,19 +101,21 @@ func (s *HTTPServer) nodeToggleDrain(resp http.ResponseWriter, req *http.Request
...
 
-	// Get the enable value
-	enableRaw := req.URL.Query().Get("enable")
-	if enableRaw == "" {
-		return nil, CodedError(400, "missing enable value")
-	}
-	enable, err := strconv.ParseBool(enableRaw)
-	if err != nil {
-		return nil, CodedError(400, "invalid enable value")
+	var drainRequest api.NodeUpdateDrainRequest
```
Comment (schmichael) [inline command/agent/node_endpoint.go]: This breaks backward compat by not supporting `enable` anymore. We could maintain it just by keeping the old query param parsing code.

File: command/node_drain.go [hunk 2]
```diff
@@ -32,8 +40,24 @@ Node Drain Options:
...
+  -force
+    Force remove allocations off the node immediately.
+
+  -no-deadline
+    No deadline allows the allocations to drain off the node without being force
+    stopped after a certain deadline.
+
+  -ignore-system
+    Ignore system allows the drain to complete without stopping system job
+    allocations.
```
Comment (schmichael) [inline command/node_drain.go]: > By default system jobs are stopped last.

File: nomad/structs/structs.go [hunk 3]
```diff
@@ -2959,10 +2977,10 @@ func (tg *TaskGroup) Validate(j *Job) error {
 	// Validate the migration strategy
 	switch j.Type {
 	case JobTypeService:
-		if tg.Count == 1 && tg.Migrate != nil {
-			mErr.Errors = append(mErr.Errors, fmt.Errorf("Task Group %v should not have a migration strategy with a count = 1", tg.Name))
-		} else if err := tg.Migrate.Validate(); err != nil {
-			mErr.Errors = append(mErr.Errors, err)
+		if tg.Migrate != nil {
```
Comment (schmichael) [inline nomad/structs/structs.go]: Why did you remove this?

File: command/agent/node_endpoint.go [hunk 4]
```diff
@@ -101,19 +101,21 @@ func (s *HTTPServer) nodeToggleDrain(resp http.ResponseWriter, req *http.Request
...
+	if err := decodeBody(req, &drainRequest); err != nil {
+		return nil, CodedError(400, err.Error())
 	}
 
 	args := structs.NodeUpdateDrainRequest{
 		NodeID: nodeID,
```
Comment (schmichael) [inline command/agent/node_endpoint.go]: Always set `Node.Drain` for backward compat?

File: nomad/structs/structs.go (hunk 3 above)
Comment (dadgar) [reply nomad/structs/structs.go]: I don't like the UX of it as user scales up and down a count they will get hit by this.

Comment (schmichael) [reply nomad/structs/structs.go]: I guess it doesn't matter since MaxParallel has to be >0, so Count==1 should Just Work whether or not we special case it.
